File: 0474-ones-and-zeroes/0474-ones-and-zeroes.cpp
class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        vector<pair<int, int>> freqs;
        vector<vector<vector<int>>> dp(n + 1, vector<vector<int>>(m + 1, vector<int>(strs.size(), -1)));
        for(int i = 0; i < strs.size(); i++)
        {
            int o = 0;
            int z = 0;
            for(int j = 0; j < strs[i].length(); j++)
            {
                if(strs[i][j] == '1')
                {
                    o++;
                }
                else
                {
                    z++;
                }
            }
            freqs.push_back({o, z});
        }
        return solve(freqs, dp, 0, 0, 0, n, m);
    }
private:
    int solve(vector<pair<int, int>>& freqs, vector<vector<vector<int>>>& dp, int i, int o, int z, int n, int m)
    {
        if(i == freqs.size())
        {
            return 0;
        }
        if(dp[o][z][i] == -1)
        {
            dp[o][z][i] = 0;
            if(o + freqs[i].first <= n && z + freqs[i].second <= m)
            {
                dp[o][z][i] = max(dp[o][z][i], solve(freqs, dp, i + 1, o + freqs[i].first, z + freqs[i].second, n, m) + 1);
            }
            dp[o][z][i] = max(dp[o][z][i], solve(freqs, dp, i + 1, o, z, n, m));
        }
        return dp[o][z][i];
    }
};

File: 0474-ones-and-zeroes/0474-ones-and-zeroes.py
import numpy as np
class Solution:
    def solve(self, freqs: List[tuple[int, int]], dp: List[List[List[int]]], i: int, o: int, z: int, n: int, m: int) -> int:
        if i == len(freqs):
            return 0
        if dp[o][z][i] == -1:
            dp[o][z][i] = 0
            if o + freqs[i][0] <= n and z + freqs[i][1] <= m:
                dp[o][z][i] = max(dp[o][z][i], self.solve(freqs, dp, i + 1, o + freqs[i][0], z + freqs[i][1], n, m) + 1)
            dp[o][z][i] = max(dp[o][z][i], self.solve(freqs, dp, i + 1, o, z, n, m))
        return int(dp[o][z][i])
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        freqs = []
        dp = np.full((n + 1, m + 1, len(strs)), -1, dtype=int)
        for s in strs:
            o = 0
            z = 0
            for ch in s:
                if ch == '1':
                    o += 1
                else:
                    z += 1
            freqs.append((o, z))
        
        return self.solve(freqs, dp, 0, 0, 0, n, m)